### PR TITLE
Update fit browser when curves added/removed

### LINKED ISSUE #31661: Workspace and index fields of fit browser not updating when curve added or removed
### Expected behavior
When adding a curve to a plot with an open fit browser the user would be able to select that curve to fit.

### Actual behavior
When a spectrum/workspace is added or removed from the plot while a fit browser is open, the workspace index and workspace drop-down box aren't refreshed. If you remove a curve from the plot settings you can still select it to fit and this causes a sad mantid/unexpected exception.

```
Traceback (most recent call last):
  File "C:\MantidInstall\bin\mantidqt\widgets\fitpropertybrowser\fitpropertybrowser.py", line 120, in _set_normalise_data
    should_normalise_before_fit = ws_artist.is_normalized and not ws_is_distribution
AttributeError: 'NoneType' object has no attribute 'is_normalized'
```

### Steps to reproduce the behavior
(1) Plot the first spectrum from e.g. HRP39182.raw (i.e. workspace index = 0)
(2) Open the fit browser
(3) Plot the second spectrum (i.e. workspace index = 1)
(4) You will not be able to change the workspace index to 1 in the fit browser settings group
(5) Toggle the fit browser off/on - now change the index to 1 (it should be possible now)
(6) Go to the plot settings > curves > remove the second spectrum (index = 1)
(7) Fit a function, you will get an exception

Or you could plot data from two workspaces and recreate the error in a similar way

### Platforms affected
Found during smoke testing fitting on windows #31624

### PR DESCRIPTION
**Pull request by StephenSmith25**
**Description of work.**
This PR ensures the workspace index spinbox in the fit browser update when curves are added / removed.

**To test:**
(1) Plot the first spectrum from some data
(2) Open the fit browser
(3) Plot the second spectrum (i.e. workspace index = 1)
(4) You should now be able to select the new spectrum
(6) Go to the plot settings > curves > remove the second spectrum (index = 1)
(7) You should no longer be able to select that spectrum in the spinbox

Fixes #31661  

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: qt/widgets/common/src/FitPropertyBrowser.cpp [hunk 1]
```diff
@@ -2451,10 +2453,6 @@ int FitPropertyBrowser::getAllowedIndex(int currentIndex) const {
     return 0;
   }
 
-  if (currentIndex == m_oldWorkspaceIndex) {
```
Comment (StephenSmith25) [inline qt/widgets/common/src/FitPropertyBrowser.cpp:2454]: I have no idea what this check was doing or why it was here :thinking:

**Comment by StephenSmith25** [comment]
@RichardWaiteSTFC When you have time could you test this. So many edge cases with things like this, hopefully this covers them and doesn't introduce any new issues..

File: qt/python/mantidqt/widgets/fitpropertybrowser/fitpropertybrowser.py [hunk 2]
```diff
@@ -80,10 +113,17 @@ def _get_allowed_spectra(self):
         for ax in self.canvas.figure.get_axes():
             try:
                 for ws_name, artists in ax.tracked_workspaces.items():
+                    # we don't want to include the fit workspace in our selection
+                    # unfortunately, this might catch some workspaces which have _Workspace appended to their name
+                    # hopefully this is an unlikely scenario.
+                    if ws_name.endswith("_Workspace"):
```
Comment (RichardWaiteSTFC) [inline qt/python/mantidqt/widgets/fitpropertybrowser/fitpropertybrowser.py:119]: I think there's a more elegant way of doing this that I implemented in the EngDiff UI when overloading  `_get_allowed_spectra` (which I think it a natural place to put this) that will discriminate between output workspaces from the fit and any workspace ending with `_Workspace` 
https://github.com/mantidproject/mantid/blob/dec8b0f0596373b906024f13c882f6881016728e/scripts/Engineering/gui/engineering_diffraction/tabs/fitting/plotting/EngDiff_fitpropertybrowser.py#L107

File: qt/widgets/common/src/FitPropertyBrowser.cpp (hunk 1 above)
Comment (RichardWaiteSTFC) [reply qt/widgets/common/src/FitPropertyBrowser.cpp:2454]: Me either perhaps it was to catch an invalid index (i.e. <0) which used to occur when this condition was met? Removing it doesn't seem to have done any harm...

**Review by RichardWaiteSTFC — CHANGES_REQUESTED** [review]
This works very nicely! The spin box now takes you to the next valid workspace index. It updates when curves are added/removed (when removing the current workspace index it will select another valid one automatically) and handles renaming and multiple workspaces etc.
I have made one comment about how the output workspaces are excluded - but maybe I have misunderstood?

File: qt/python/mantidqt/widgets/fitpropertybrowser/fitpropertybrowser.py (hunk 2 above)
Comment (RichardWaiteSTFC) [reply qt/python/mantidqt/widgets/fitpropertybrowser/fitpropertybrowser.py:119]: So `output_wsnames` will include all the `_Workspace`, `_Parameters `etc. workspaces - though I'm not sure they are updated if e.g. the user was to rename one of the output workspaces (i.e. I think whatever is returned by `self.getWorkspaceList()` is updated at the point of the fit completion, it is not a dynamically updated thing)

Comment (StephenSmith25) [reply qt/python/mantidqt/widgets/fitpropertybrowser/fitpropertybrowser.py]: Thanks for pointing me towards that - looks like a more robust approach

**Comment by StephenSmith25** [comment]
> This works very nicely! The spin box now takes you to the next valid workspace index. It updates when curves are added/removed (when removing the current workspace index it will select another valid one automatically) and handles renaming and multiple workspaces etc.
> I have made one comment about how the output workspaces are excluded - but maybe I have misunderstood?

Thanks for testing that all!

**Review by RichardWaiteSTFC — DISMISSED** [review]
Looks good, nice fix and thanks for the change!

File: qt/python/mantidqt/widgets/test/test_fitpropertybrowser.py [hunk 3]
```diff
@@ -32,7 +31,10 @@ def tearDown(self):
         AnalysisDataService.clear()
 
     def test_initialization_does_not_raise(self):
-        assertRaisesNothing(self, self._create_widget)
+        try:
```
Comment (DavidFair) [inline qt/python/mantidqt/widgets/test/test_fitpropertybrowser.py:34]: The correct way of testing this is:
```python
self.assertIsNone(self._createWidget())
```
Or alternatively
```python
self._createWidget()
```

All unit tests assume that exceptions are failures so the try-expect is redundant and will likely set off future static analysis about the bare except

File: qt/python/mantidqt/widgets/test/test_fitpropertybrowser.py [hunk 4]
```diff
@@ -110,37 +135,13 @@ def test_fit_result_workspaces_are_added_to_browser_when_fitting_done(self):
...
-        # click on matrix workspace
-        MatrixWorkspaceDisplay.show_view = Mock()
-        item = wsList.item(0).text()
-        property_browser.workspaceClicked.emit(item)
-        self.assertEqual(1, MatrixWorkspaceDisplay.show_view.call_count)
-
-    def test_fit_parameter_table_workspaces_in_browser_is_viewed_when_clicked(self):
         if not PYQT5:
             self.skipTest("MatrixWorkspaceDisplay and TableWorkspaceDisplay cannot be "
                           "imported in qt4 so the test fails with an error.")
```
Comment (DavidFair) [inline qt/python/mantidqt/widgets/test/test_fitpropertybrowser.py:140]: This is dead code and can be removed, especially as it will incorrectly skip the test if we go to PyQt6

File: qt/widgets/common/src/FitPropertyBrowser.cpp [hunk 5]
```diff
@@ -3261,7 +3259,16 @@ void FitPropertyBrowser::addAllowedSpectra(const QString &wsName, const QList<in
     for (auto const i : wsSpectra) {
       indices.push_back(static_cast<int>(ws->getIndexFromSpectrumNumber(i)));
     }
+    auto wsFound = m_allowedSpectra.find(wsName);
     m_allowedSpectra.insert(wsName, indices);
+    if (wsFound != m_allowedSpectra.end()) {
```
Comment (DavidFair) [inline qt/widgets/common/src/FitPropertyBrowser.cpp:3264]: This doesn't need to be done now, but just a heads-up for the future we can write this all in the if block

```c++
if (auto wsFound = m_allowedSpectra.find(wsName); wsFound!=m_allowedSpectra.end()) {

**Review by DavidFair — CHANGES_REQUESTED** [review]
Some minor stuff on the unit tests which could cause headaches in the future 
Otherwise looks good

File: qt/python/mantidqt/widgets/test/test_fitpropertybrowser.py (hunk 3 above)
Comment (StephenSmith25) [reply qt/python/mantidqt/widgets/test/test_fitpropertybrowser.py]: But create_widget does return an object, so `assertIsNone` would always fail surely. To best honest this test is redundant as you say, as the method is called in every subsequent test, so we'll always know about it. I've removed it.

**Review by DavidFair — APPROVED** [review]
Thanks :+1:
